transports/webrtc/src/transport.rs

```diff
@@ -0,0 +1,697 @@
+                Poll::Pending => Poll::Pending,
+            };
+        }
+    }
+}
+
+impl WebRTCTransport {
+    async fn do_dial(self, addr: Multiaddr) -> Result<Connection<'static>, Error> {
+        let socket_addr =
+            multiaddr_to_socketaddr(&addr).ok_or_else(|| Error::InvalidMultiaddr(addr.clone()))?;
+        if socket_addr.port() == 0 || socket_addr.ip().is_unspecified() {
+            return Err(Error::InvalidMultiaddr(addr.clone()));
+        }
+
+        let config = self.config.clone();
+
+        let remote = addr.clone(); // used for logging
+        trace!("dialing address: {:?}", remote);
+
+        let fingerprint = self.cert_fingerprint();
+        let se = build_setting_engine(self.udp_mux.clone(), &socket_addr, &fingerprint);
+        let api = APIBuilder::new().with_setting_engine(se).build();
+
+        let peer_connection = api
+            .new_peer_connection(config)
+            .map_err(Error::WebRTC)
+            .await?;
+
+        // Create a datachannel with label 'data'
+        let data_channel = peer_connection
+            .create_data_channel(
+                "data",
+                Some(RTCDataChannelInit {
+                    negotiated: None,
+                    id: Some(1),
+                    ordered: None,
+                    max_retransmits: None,
+                    max_packet_life_time: None,
+                    protocol: None,
+                }),
+            )
+            .await?;
+
+        let (data_channel_rx, data_channel_tx) = oneshot::channel::<Arc<DetachedDataChannel>>();
+
+        // Wait until the data channel is opened and detach it.
+        let d = Arc::clone(&data_channel);
+        data_channel
+            .on_open(Box::new(move || {
+                debug!("Data channel '{}'-'{}' open.", d.label(), d.id());
+
+                let d2 = Arc::clone(&d);
+                Box::pin(async move {
+                    match d2.detach().await {
+                        Ok(detached) => {
+                            if let Err(_) = data_channel_rx.send(detached) {
+                                error!("data_channel_tx dropped");
+                            }
+                        },
+                        Err(e) => {
+                            error!("Can't detach data channel: {}", e);
+                        },
+                    };
+                })
+            }))
+            .await;
+
+        let offer = peer_connection
+            .create_offer(None)
+            .map_err(Error::WebRTC)
+            .await?;
+        debug!("OFFER: {:?}", offer.sdp);
+        peer_connection
+            .set_local_description(offer)
+            .map_err(Error::WebRTC)
+            .await?;
+
+        // Set the remote description to the predefined SDP.
+        let fingerprint = match addr.iter().last() {
+            Some(Protocol::XWebRTC(f)) => f,
+            _ => {
+                return Err(Error::InvalidMultiaddr(addr));
+            },
+        };
```

Comment:
Can use `ok_or` + `?` here.

Reply:
no because I'm matching a specific protocol (XWebRTC)

Reply:
I think you should be able to map it with `and_then`: https://doc.rust-lang.org/std/option/enum.Option.html#method.and_then

However, at this point, what you have now is likely easier to read :)